src/thermal_conductivity/scattering_threephonon.f90

```diff
@@ -0,0 +1,395 @@
+            perm = 1.0_r8
+        else
+            perm = 2.0_r8
+        end if
+
+        ! This get the ifc3 in Fourier space, but not on phonons
+        call pretransform_phi3(fct, qp%ap(q2)%r, qp%ap(q3)%r, ptf)
+        do b2 = 1, dr%n_mode
+            om2 = dr%aq(q2)%omega(b2)
+            if (om2 .lt. lo_freqtol) cycle
+
+            egv2 = dr%aq(q2)%egv(:, b2)/sqrt(om2)
+
+            ! This is the multiplication of eigv of phonons 1 and 2
+            evp1 = 0.0_r8
+            call zgeru(dr%n_mode, dr%n_mode, (1.0_r8, 0.0_r8), egv2, 1, egv1, 1, evp1, dr%n_mode)
+            do b3 = 1, dr%n_mode
+                om3 = dr%aq(q3)%omega(b3)
+                if (om3 .lt. lo_freqtol) cycle
+                egv3 = dr%aq(q3)%egv(:, b3)/sqrt(om3)
+
+                select case (integrationtype)
+                case (1)
+                    sigma = smearing*lo_frequency_THz_to_Hartree
+                case (2)
+                    sigma = sqrt(sr%sigsq(q1, b1) + &
+                                 sr%sigsq(qp%ap(q2)%irreducible_index, b2) + &
+                                 sr%sigsq(qp%ap(q3)%irreducible_index, b3))
+                end select
+
+                ! This is the multiplication of eigv of phonons 1 and 2 and now 3
+                evp2 = 0.0_r8
+                call zgeru(dr%n_mode, dr%n_mode**2, (1.0_r8, 0.0_r8), egv3, 1, evp1, 1, evp2, dr%n_mode)
+                evp2 = conjg(evp2)
+                ! And with this, we have the scattering matrix element coming in
+                c0 = dot_product(evp2, ptf)
+                psisq = threephonon_prefactor*abs(c0*conjg(c0))*mcg%weight
+
+                ! Let's get the Bose-Einstein distributions
+                n2 = sr%be(qp%ap(q2)%irreducible_index, b2)
+                n3 = sr%be(qp%ap(q3)%irreducible_index, b3)
+
+                ! The prefactor for the scattering
+                plf0 = n2 - n3
+                plf1 = n2 + n3 + 1.0_r8
+                f0 = perm*psisq*plf0*(lo_gauss(om1, -om2 + om3, sigma) - lo_gauss(om1, om2 - om3, sigma))
+                f1 = perm*psisq*plf1*(lo_gauss(om1, om2 + om3, sigma) - lo_gauss(om1, -om2 - om3, sigma))
+
+                ! And we add everything for each triplet equivalent to the one we are actually computing
+                do i = 1, size(red_triplet, 2)
+                    ! We accumulate the linewidth
+                    g0 = g0 + f0 + f1
+
+                    ! And also the off-diagonal part of the scattering matrix
+                    q2p = red_triplet(1, i)
+                    q3p = red_triplet(2, i)
+                    od_terms(q2p, b2) = od_terms(q2p, b2) + 2.0_r8*(f0 + f1)*om2/om1
+                    od_terms(q3p, b3) = od_terms(q3p, b3) + 2.0_r8*(f0 + f1)*om3/om1
+                end do
+            end do
+        end do
+    end do compute_loop
+
+    ! Now we can symmetrize the off-diagonal contribution
+    ! This can be done in a way to put a value to for mode that have been skipped by the Monte-Carlo !
```

Comment:
Not my best comment, but I think that is an important point since it's easy to miss. I will try to make it clearer.
But the idea is that when I symmetrize the scattering matrix, when I average over the equivalent entries, I sum over non-zero entry - because q-point not in the Monte-Carlo grid will be zero at this point. But when I distribute the average values, I distribute on all equivalent point, even the one not in the Monte-Carlo grid